src/cheri_pte.sail

```diff
@@ -38,102 +38,107 @@ union PTE_Check = {
-                                         }
+                                         (true, ext_ptw_join_lcm(ext_ptw,
+                                                  if   e.LoadCap() == 0b1
+                                                  then PTW_LCM_OK
+                                                  else /*
+                                                        * Allow the address translation to proceed,
+                                                        * but mark for tag stripping
+                                                        */
+                                                       PTW_LCM_CLR))
                                        }
-                                       else (false, PTE_CAP_OK)
+                                       else (false, ext_ptw)
                                      },
 
     (Write(Cap),      User)       => { if   p.U() == 0b1 & p.W() == 0b1
                                        then {
                                          let  e = Mk_Ext_PTE_Bits(ext);
                                          if   e.StoreCap() == 0b1
-                                         then (true, PTE_CAP_OK)
+                                         then (true, ext_ptw)
                                          else {
                                            /* Do not allow the address translation to proceed */
-                                           (false, PTE_STORE_CAP_ERR)
+                                           (false, ext_ptw_join_scm(ext_ptw, PTW_SCM_TRAP))
                                          }
                                        }
-                                       else (false, PTE_CAP_OK)
+                                       else (false, ext_ptw)
                                      },
 
     (ReadWrite(Cap),  User)       => { if   p.U() == 0b1 & p.R() == 0b1 & p.W() == 0b1
                                        then {
-                                         let  e = Mk_Ext_PTE_Bits(ext);
-                                         if   e.StoreCap() == 0b1 & e.LoadCap() == 0b1
-                                         then (true, PTE_CAP_OK)
-                                         else if e.StoreCap() == 0b0
-                                         /* return a failure since we should cause an exception */
-                                         then (false, PTE_STORE_CAP_ERR)
-                                         /* return a success since the translation should proceed */
-                                         else (true, PTE_LOAD_CAP_ERR)
+                                         let    e = Mk_Ext_PTE_Bits(ext);
+                                         let oksc = e.StoreCap() == 0b1;
+                                         let oklc = e.LoadCap() == 0b1;
+                                         ( /* Translation succeeds if SC also OK */
+                                           oksc,
```

Comment:
Isn't this wrong? Translation succeeds, you just need to do a data-dependent trap. Similarly with the plain Write and Supervisor cases.

Reply:
I was just following along the existing code for this patch, but yes, that's probably the intended behaviour.  (We could also get it by making the request type -- `Write(Data)` vs `Write(Cap)` -- depend on the tag, but since we're going to be expressing data-dependence as post-TLB for other things, consistency is probably a grand plan.)